### PR TITLE
BLAKE2 refactoring of common code bringing variable length support to BLAKE2s

### PR DESCRIPTION
**Pull request by nabijaczleweli**
The first three patches are functional no-ops, and just move the BLAKE2B stuff from being declared directly into a macro.

The fourth patch replaces the manual implementation of BLAKE2s with a call to the macro, thus matching them both in funxionality. This effectively completes 786b9a8d3f8e203c5536e36b9a9bab83bde0311a.

The fifth patch explicitly documents what I discovered the hard way over three hours yesterday:
* openssl has `EVP_DigestFinalXOF()` which is a shorthand for setting "xoflen" to the argument and then running `EVP_DigestFinal()`
* this matches how SHAKE works (output length has no bearing on the hashing, so you can set "xoflen" at any time)
* BLAKE2 puts the output length into the parameter block, which is part of the seed
* naturally, you need to pass the output length to the init, so you set "xoflen" in the params block for `EVP_DigestInit_ex2()`
* setting it after init and before the first byte does nothing, but you can still do it
* `EVP_DigestFinalXOF()` has no bearing on it at all, but you can still pass it there. or you can just use `EVP_DigestFinal()`
* if you didn't set it in `EVP_DigestInit_ex2()` but did pass it to `EVP_DigestFinalXOF()`, you just get wrong data (full-width hash, but truncated)
* support for "xoflen" in BLAKE2b appeared in 786b9a8d3f8e203c5536e36b9a9bab83bde0311a in 2021, and is only in openssl-3.2.0-alpha[12] tags
* `EVP_DigestInit_ex2()` *doesn't error* if it gets a parameter the hash function doesn't understand
* I know this because on Bookworm openssl 3.0.11, code with full error-checking and `EVP_DigestFinal()` produces a truncated hash, and on upstream openssl HEAD the same code produces the correct short hash
* you only get an error if you use `EVP_DigestFinalXOF()` which you shouldn't be using since it does nothing but implies it does

This all could've been avoided with a startlingly small patch that documents it (+20-4, and most of that +20 is in EVP_MD-BLAKE2.pod which is mostly copied from -SHAKE). The first four patches are really in service of not having to differentiate between b2s and b2b in the manual.

This also means that `EVP_DigestInit_ex2()` blindly accepting whatever it doesn't understand is an oddly dangerous design?

### CODE REVIEW

**Comment by t8m** [comment]
@mattcaswell I actually think this should go in 3.2.0

The refactoring is fairly straightforward and the xoflen support in 2b only is quite surprising.

The documentation is clearly needed as the xoflen handling in Blake is very much different from SHAKE.

**Review by t8m — DISMISSED** [review]

**Review by tmshort — DISMISSED** [review]
LGTM

**Comment by paulidale** [comment]
@slontis, does this impact your proposal for the digest squeeze function?

File: doc/man7/EVP_MD-BLAKE2.pod [hunk 1]
```diff
@@ -30,6 +30,23 @@ Known names are "BLAKE2B-512" and "BLAKE2b512".
...
 in L<EVP_MD-common(7)>.
 
+=head2 Settable Context Parameters
+
+These implementations support the following L<OSSL_PARAM(3)> entries,
+settable for an B<EVP_MD_CTX> with L<EVP_MD_CTX_set_params(3)>:
+
+=over 4
+
+=item "xoflen" (B<OSSL_DIGEST_PARAM_XOFLEN>) <unsigned integer>
```
Comment (paulidale) [inline doc/man7/EVP_MD-BLAKE2.pod:40]: I am not convinced that `OSSL_DIGEST_PARAM_XOFLEN>` (_xoflen_) is the correct parameter here because there is no capability for _extension_ of the output.  I.e. it's not really an XOF digest despite having the capability for variable output length.

To me, OSSL_DIGEST_PARAM_SIZE` (_size_) seems more appropriate.

There is work in progress to add proper XOF support which won't work for these.  The [design](https://github.com/openssl/openssl/pull/21443) and [implementation](https://github.com/openssl/openssl/pull/21511) PRs are outstanding.

**Review by paulidale — CHANGES_REQUESTED** [review]

**Comment by paulidale** [comment]
OTC: **should we classify these are XOF and use _xoflen_ for the digest's output size?  The alternative is that they are not XOF and use _size_.**

File: doc/man7/EVP_MD-BLAKE2.pod (hunk 1 above)
Comment (t8m) [reply doc/man7/EVP_MD-BLAKE2.pod:40]: Using OSSL_DIGEST_PARAM_SIZE as settable parameter would be quite weird. And although Blake2 is not really XOF the EVP_DigestFinalXOF is not really XOF function as there is nothing extensible in the current implementation of it and even Shane's PRs linked above do not change anything about that.

The problem is Blake2b is already made variable-length output function in the current master branch but Blake2s isn't which is confusing.

**Comment by Tarnum-tst** [comment]
This PR conflicts with 6a0ae393dd554eb718e5148696e8f437d4faae5b
So needs rebasing/fixing.

**Comment by nabijaczleweli** [comment]
rebased

**Comment by t8m** [comment]
@paulidale AFAIK the OTC hold question is not relevant anymore. Can you please remove the label if you agree?

**Comment by t8m** [comment]
@nabijaczleweli I think you'll also need to modify EVP_MD-Blake2 manpage and I'd kindly ask you to add a CHANGES.md entry about adding the variable size support to Blake2s.

**Comment by nabijaczleweli** [comment]
Well, if we're going that far, I should probably document that BLAKE2b also has it? 786b9a8 doesn't have any docs in it, incl. CHANGES.md.

**Comment by paulidale** [comment]
Hold removed, we should have removed it after the discussion.

For the record, the OTC decided that this wasn't an XOF digest and that the digest size parameter should be used to specify the length.  This change to be made to the pre-existing digest before 3.2 beta series releases (because the variable length change hadn't been in a release, it could be updated so long as it was before the API freeze at the first 3.2 beta release).

**Comment by nabijaczleweli** [comment]
EVP_MD-BLAKE2(7) was already modified with text much like was in this PR in https://github.com/openssl/openssl/commit/6a0ae393dd554eb718e5148696e8f437d4faae5b#diff-a94d99d3c13153981b6da8a09b5d5fefcb3f8d6016c8cf0752f44fb30addc919.

CHANGES.md entry added.

File: CHANGES.md [hunk 2]
```diff
@@ -33,6 +33,11 @@ OpenSSL 3.2
 
 ### Changes between 3.1 and 3.2 [xx XXX xxxx]
 
+ * The BLAKE2b and BLAKE2s hash algorithms support a configurable output length
+   by setting the "size" parameter.
+
+   *Čestmír Kalina, Tomáš Mráz, and Ahelenia Ziemiańska*
+
```
Comment (t8m) [inline CHANGES.md:40]: Unfortunately this needs two entries. One in `### Changes between 3.1 and 3.2` for BLAKE2b and another in changes between 3.2 and 3.3 for BLAKE2s. Also the first one needs to be in a separate PR so it gets merged to the 3.2 branch as well.

Comment (nabijaczleweli) [reply CHANGES.md:40]: does it? I only see 786b9a8d3f8e203c5536e36b9a9bab83bde0311a in `openssl-3.2.0-{alpha{1,2},beta1}` tags, and it doesn't look backported into `openssl-3.1.4` under another SHA either?

Comment (nabijaczleweli) [reply CHANGES.md]: yep, misread and misunderstood. fixed

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

File: providers/implementations/digests/blake2_prov.c [hunk 3]
```diff
@@ -8,123 +8,181 @@
...
+    if (params == NULL) \
+        return 1; \
+ \
+    p = OSSL_PARAM_locate_const(params, OSSL_DIGEST_PARAM_SIZE); \
+    if (p != NULL) { \
+        if (!OSSL_PARAM_get_size_t(p, &size)) { \
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER); \
+            return 0; \
+        } \
+        if (size < 1 || size > UINT8_MAX) { \
```
Comment (slontis) [inline providers/implementations/digests/blake2_prov.c:77]: Is this upper constraint correct.. RFC 7693 seems to have 1 <= outlen <=64

Comment (nabijaczleweli) [reply providers/implementations/digests/blake2_prov.c:77]: libb2 also refuses lengths longer than the hash, but the paper says "put 0xFFFFFFFF in there if you don't know" () – I daren't have a take on this, I'm just adding backslashes.

File: test/recipes/30-test_evp_data/evpmd_blake.txt [hunk 4]
```diff
@@ -54,6 +54,16 @@ Digest = BLAKE2s256
...
+Digest = BLAKE2s256
+Input =
+OutputSize = 16
+Output = 64550d6ffe2c0a01a14aba1eade0200c
+
+Digest = BLAKE2s256
+Input = 61
+OutputSize = 10
+Output = b60d322755eebca92b5e
+
```
Comment (slontis) [inline test/recipes/30-test_evp_data/evpmd_blake.txt:66]: If there is an upper bound on the output length there should be failure checks here..

Comment (nabijaczleweli) [reply test/recipes/30-test_evp_data/evpmd_blake.txt:66]: Similarly, this just mimicks the blake2b tests at the end of the file (but obviously with blake2s outputs). If the limit in the formerly-blake2b code is appropriate, then maybe. No limit is documented afaict

**Comment by t8m** [comment]
Merged to the master branch with partially squashing the commits. Thank you for your contribution.
